### PR TITLE
[PSYS-123] Upgrade for PHP 8.1 Compatibility

### PR DESCRIPTION
**Pull request by sjohnston-digistorm**
### Fix PSYS-123
Link to ticket: https://digistorm.atlassian.net/browse/PSYS-123

✅ PHPUnit tests pass
✅ Payments tests against this branch of the package pass
- Applied Rector PHP 8.1 rules
- Converted the enum classes to enums
- Introduced types
- Added Carbon for date handling
- Added some more tests

Once merged, will push a new major version (3.0.0).

### CODE REVIEW

File: src/CardProxy.php [hunk 1]
```diff
@@ -10,104 +15,70 @@
...
     /**
-     * @return CardProxy
+     * @throws SchoolEasyPayException
      */
-    public function setExpiry(mixed $expiry)
+    public function setExpiry(string $expiry): self
     {
-        $this->expiry = $expiry;
-
+        $dateStringWithDay = '01/' . $expiry;
```
Comment (zleigh-digistorm) [inline src/CardProxy.php:68]: What is this doing?

Comment (sjohnston-digistorm) [reply src/CardProxy.php:68]: Expiry is provided in the format mm/yy.  To turn it into a carbon date object, a day of the month (in this case first of the month) is prepended to the string.  Then on the get, the carbon date is converted back to a string with the format 'mm/yy'.

File: src/Traits/HasCarbonDates.php [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+use Carbon\Carbon;
+use Digistorm\SchoolEasyPay\Exceptions\SchoolEasyPayException;
+
+trait HasCarbonDates
+{
+    private function castCarbonToString(Carbon $date, string $outputFormat = 'Y-m-d\TH:i:s.v\Z'): string
+    {
+        return $date->format($outputFormat);
+    }
+    /**
```
Comment (zleigh-digistorm) [inline src/Traits/HasCarbonDates.php:16]: Should have a new line here

Comment (sjohnston-digistorm) [reply src/Traits/HasCarbonDates.php:16]: Thanks, updated

File: tests/Unit/CustomerTest.php [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    {
+        $customer = new Customer($this->config);
+        $customerReference = 'REF12345';
+        $customer->setCustomerReference($customerReference);
+
+        $this->assertEquals($customerReference, $customer->getCustomerReference());
+    }
+
+    /**
+     * @throws SchoolEasyPayException
```
Comment (zleigh-digistorm) [inline tests/Unit/CustomerTest.php]: Maybe we don't need these comments in the tests?

Comment (sjohnston-digistorm) [reply tests/Unit/CustomerTest.php]: I just added them to satisfy the IDE (it complains with a warning / yellow line otherwise).

Comment (sjohnston-digistorm) [reply tests/Unit/CustomerTest.php]: I've wrapped the relevant calls in `try` / `catch` to avoid needing the annotation.

**Review by zleigh-digistorm — APPROVED** [review]

**Review by mlangford-digistorm — APPROVED** [review]

**Pull request merged**